test/unified-config.test.js

```diff
@@ -0,0 +1,453 @@
+
+  //============================
+  // individual tests start here
+  //============================
+
+  it('should default correctly when no config file or env vars', function () {
+    const cfg = guc();
+
+    doChecks(cfg);
+  })
+
+  it('should set known config keys in a config file', function () {
+    const config = {
+      enabled: false,
+      hostnameAlias: 'bruce',
+      serviceKey: 'f'.repeat(64),
+      ignoreConflicts: true,
+      domainPrefix: 'fubar',
+    }
+    writeConfigJSON(config);
+
+    const cfg = guc();
+
+    doChecks(cfg, {global: config});
+  })
+
+  it('should warn about deprecated config file keys', function () {
+    const config = {
+      traceMode: 'always',
+      sampleRate: 1000000,
+    }
+    writeConfigJSON(config);
+
+    const cfg = guc();
+
+    // fix up to match translated value.
+    config.traceMode = 1;
+    const warnings = [
+      'traceMode is deprecated; it will be invalid in the future',
+      'sampleRate is deprecated; it will be invalid in the future',
+    ];
+    doChecks(cfg, {global: config, warnings});
+  })
+
+  it('should warn about unknown config keys', function () {
+    const config = {badKeyRising: 'i have been a bad key'};
+    writeConfigJSON(config);
+
+    const cfg = guc();
+
+    const unusedConfig = ['badKeyRising: i have been a bad key'];
+    doChecks(cfg, {unusedConfig});
+  })
+
+  it('should not set keys with invalid values', function () {
+    const config = {ec2MetadataTimeout: 'hello', createTraceIdsToken: 'bruce'};
+    writeConfigJSON(config);
+
+    const cfg = guc();
+
+    const warnings = [
+      'invalid configuration file value createTraceIdsToken: bruce',
+      'invalid configuration file value ec2MetadataTimeout: hello',
+    ];
+    doChecks(cfg, {warnings});
+  })
+
+  it('should handle a non-default config file correctly', function () {
+    const config = {ec2MetadataTimeout: 4000};
+    const filename = 'non-default.json';
+    fs.writeFileSync(filename, JSON.stringify(config));
+
+    process.env.APPOPTICS_APM_CONFIG_NODE = filename;
+
+    const cfg = guc();
+
+    fs.unlinkSync(filename);
+
+    doChecks(cfg, {file: `${process.cwd()}/${filename}`, global: config});
+  })
+
+  //
+  // config file errors
+  //
+  it('should correctly report an error reading a non-default config file', function () {
+    const file = process.env.APPOPTICS_APM_CONFIG_NODE = 'xyzzy-bruce.json';
+    const fullpath = `${process.cwd()}/${file}`;
+
+    const cfg = guc();
+
+    const message = 'Cannot find module \'/home/bruce/solarwinds/ao-agent/xyzzy-bruce.json\'';
+    const errors = [
+      `Cannot read config file ${fullpath}: ${message}`
+    ]
+    doChecks(cfg, {file: fullpath, errors});
+  })
+
+  it('should correctly report an error reading a bad config file', function () {
+    writeConfigLiteral('{"i am: bad\n');
+
+    const cfg = guc();
+
+    const errors = [[
+      'Cannot read config file /home/bruce/solarwinds/ao-agent/appoptics-apm: ',
```

Comment:
there is a duplication here and I think the filename should have the extension '.js'
Same issue with the path as previous test.

Reply:
i don't follow the duplication. what do you mean?

the filename shouldn't have the extension `.js` because the message is echoing the default name used. that does not specify an extension (so it will find `.js` files before `.json` files). but the later portion of the message is the error that node throws with the specific file that it was reading.